### PR TITLE
REF: pyepics shim

### LINKED ISSUE #715: New pyepics features call for refactor of pyepics_shim
* Better thread safety on `epics.PV` should obviate the need for shim-level threading locks:
https://github.com/NSLS-II/ophyd/blob/cf2b85c2479d645c144289fca79f74b151183417/ophyd/_pyepics_shim.py#L43-L45

* Automatic context switching for `epics.PV` objects should allow for the removal of:
https://github.com/NSLS-II/ophyd/blob/cf2b85c2479d645c144289fca79f74b151183417/ophyd/_pyepics_shim.py#L137

* `auto_monitor` reconfiguration is now possible with an already-created `epics.PV`. Can be refactored:
https://github.com/NSLS-II/ophyd/blob/cf2b85c2479d645c144289fca79f74b151183417/ophyd/_pyepics_shim.py#L68

### PR DESCRIPTION
**Pull request by klauer**
Builds on #710 

Closes #715 

TODO
------
* ~Remove temporary commits~
* ~Wait for pyepics version bump~
* ~Increase required version check in pyepics shim~

### CODE REVIEW

**Comment by klauer** [comment]
All pyepics tests are now passing, but the interpreter is segfaulting on Travis post-pytest. Can't reproduce locally on OSX:
```
 1581 passed, 121 skipped, 1701 deselected, 1 xpassed, 6 warnings in 255.04 seconds 
/home/travis/.travis/functions: line 104:  5565 Segmentation fault      (core dumped) coverage run --concurrency=thread --parallel-mode run_tests.py -v -k "${TEST_CL}"
The command "coverage run --concurrency=thread --parallel-mode run_tests.py -v -k "${TEST_CL}"" exited with 139.
```

Perhaps the pyepics new `auto_monitor` implementation has some bugs. Need to investigate on Linux.

File: ophyd/_pyepics_shim.py [hunk 1]
```diff
@@ -168,29 +137,32 @@ def get_pv(pvname, form='time', connect=False, context=None, timeout=5.0,
...
-    #                                             connection_callback)
-    #         thispv.connection_callbacks.append(connection_callback)
-    #     if thispv.connected:
-    #         if connection_callback:
-    #             thispv.force_connect()
-    #         if access_callback:
-    #             thispv.force_read_access_rights()
-
-    if thispv is None:
+    thispv = epics.pv._PVcache_.get((pvname, form, context))
```
Comment (danielballan) [inline ophyd/_pyepics_shim.py]: Any chance of pyepics giving us public API for this?

File: ophyd/tests/conftest.py [hunk 2]
```diff
@@ -63,7 +63,8 @@ def motor(request, cleanup):
     motor.wait_for_connection()
     motor.low_limit_value.put(-100, wait=True)
     motor.high_limit_value.put(100, wait=True)
-
+    # set the motor to 0
```
Comment (danielballan) [inline ophyd/tests/conftest.py]: :100: comment

Comment (klauer) [reply ophyd/tests/conftest.py]: Allow me to fix it:
```suggestion
    # motor.set(0)
```

File: ophyd/_pyepics_shim.py (hunk 1 above)
Comment (klauer) [reply ophyd/_pyepics_shim.py]: You mean access to the `_PVcache_`?

File: ophyd/tests/conftest.py (hunk 2 above)
Comment (tacaswell) [reply ophyd/tests/conftest.py]: leave more comments they said....

File: ophyd/_pyepics_shim.py (hunk 1 above)
Comment (klauer) [reply ophyd/_pyepics_shim.py]: Replacing the majority of `get_pv` is now possible with pyepics directly, as of https://github.com/pyepics/pyepics/pull/158 (3.4.0rc2)

File: ophyd/tests/conftest.py (hunk 2 above)
Comment (danielballan) [reply ophyd/tests/conftest.py]: More of this, please. :-D

File: ophyd/_pyepics_shim.py (hunk 1 above)
Comment (danielballan) [reply ophyd/_pyepics_shim.py]: Sorry for the slow response --- yes. Sounds good.

**Comment by tacaswell** [comment]
I get random failures and the segfault locally.

The re-use seems to fail reliably on pyepics master.

**Comment by klauer** [comment]
Force-pushed to clean out temporary commits. Tests will not pass again until the new pyepics is released, which will likely be by Monday.

**Comment by klauer** [comment]
OK, I think this is ready. @tacaswell, @danielballan, @mrakitin?

File: ophyd/_pyepics_shim.py [hunk 3]
```diff
@@ -33,16 +40,14 @@ def attach_context(self):
...
-        ca.detach_context()
+        if ca.current_context() is not None:
+            ca.detach_context()
 
 
 class PyepicsShimPV(epics.PV):
     def __init__(self, pvname, callback=None, form='time', verbose=False,
                  auto_monitor=None, count=None, connection_callback=None,
                  connection_timeout=None, access_callback=None):
-        self._get_lock = threading.Lock()
```
Comment (teddyrendahl) [inline ophyd/_pyepics_shim.py]: 🎉

File: ophyd/_pyepics_shim.py [hunk 4]
```diff
@@ -134,71 +113,20 @@ def clear_callbacks(self):
...
-            ca.write('cannot connect to %s' % pvname)
-    return thispv
+            if pv.chid is not None:
+                # Clear the channel on the CA-level
+                epics.ca.clear_channel(pv.chid)
+
+            pv.chid = None
+            pv.context = None
+
+            # Ensure we don't get this same PV back again
```
Comment (teddyrendahl) [inline ophyd/_pyepics_shim.py]: Just asking for explanation here.

Is there a bug in `pyepics` where we don't trust reconnect?

File: ophyd/tests/test_utils.py [hunk 5]
```diff
@@ -32,19 +32,6 @@ def test_waveform_to_string():
     assert epics_utils.waveform_to_string(asc) == s
 
 
-def test_pyepics_version_support():
```
Comment (teddyrendahl) [inline ophyd/tests/test_utils.py]: Kind of an unnecessary test, but did it really need to be blown up? Wouldn't changing `good_versions` to `_min_pyepics` future proof this?

Comment (tacaswell) [reply ophyd/tests/test_utils.py]: This was to test some weird strings we were getting out of pyepics 4 years ago, given that this also pins to pyepics > 3.4.0 we don't need to handle those cases anymore.

File: ophyd/_pyepics_shim.py (hunk 4 above)
Comment (tacaswell) [reply ophyd/_pyepics_shim.py]: That is a good question, it does look like pyepics is able to resuscitate a PV that has had it's channel stripped off.  There may have been something funny with the auto-monitor logic?

Seems like good house-keeping to once ophyd is done with a PV to tear it all the way down.....

File: ophyd/tests/test_utils.py (hunk 5 above)
Comment (tacaswell) [reply ophyd/tests/test_utils.py]: also, we are now using `distutils` to parse the version strings so I don't think we need to test that they are working correctly.

File: ophyd/_pyepics_shim.py (hunk 4 above)
Comment (teddyrendahl) [reply ophyd/_pyepics_shim.py]: I think going into the cache and wiping itself out will cause issues with anyone who has created `pyepics.PV` separately from the shim. I don't see this as a dealbreaker but think it should at least be mentioned.

**Comment by teddyrendahl** [comment]
👍 try saying "seems like sane shim" 3x fast

**Pull request merged**